reference-implementation/lib/writable-stream.js

```diff
@@ -182,215 +190,242 @@ function WritableStreamAddWriteRequest(stream) {
+        defaultWriterReadyPromiseReject(writer, reason);
+      } else {
+        defaultWriterReadyPromiseResetToRejected(writer, reason);
+      }
+      writer._readyPromise.catch(() => {});
+    }
+
+    WritableStreamRejectPromisesInReactionToError(stream, reason);
+
+    return;
   }
 
-  WritableStreamRejectPromisesInReactionToError(stream);
+  stream._pendingAbortRequest._reject(reason);
+  stream._pendingAbortRequest = undefined;
+
+  WritableStreamRejectPromisesInReactionToError(stream, reason);
 }
 
-function WritableStreamFinishPendingClose(stream) {
-  assert(stream._pendingCloseRequest !== undefined);
-  stream._pendingCloseRequest._resolve(undefined);
-  stream._pendingCloseRequest = undefined;
+function WritableStreamFinishInFlightClose(stream) {
+  assert(stream._inFlightCloseRequest !== undefined);
+  stream._inFlightCloseRequest._resolve(undefined);
+  stream._inFlightCloseRequest = undefined;
 
   const state = stream._state;
 
-  let wasAborted = false;
-  if (stream._pendingAbortRequest !== undefined) {
-    wasAborted = true;
-  }
-
   if (state === 'errored') {
-    if (wasAborted === true) {
+    if (stream._pendingAbortRequest !== undefined) {
       stream._pendingAbortRequest._reject(stream._storedError);
       stream._pendingAbortRequest = undefined;
     }
 
-    WritableStreamRejectClosedPromiseIfAny(stream);
+    WritableStreamRejectClosedPromiseIfAny(stream, stream._storedError);
 
     return;
   }
 
-  assert(state === 'closing');
+  assert(state === 'writable');
+
+  if (stream._pendingAbortRequest === undefined) {
+    stream._state = 'closed';
 
-  if (wasAborted === false) {
     const writer = stream._writer;
     if (writer !== undefined) {
       defaultWriterClosedPromiseResolve(writer);
     }
-    stream._state = 'closed';
+
     return;
   }
 
   stream._pendingAbortRequest._resolve();
   stream._pendingAbortRequest = undefined;
 
+  const error = new TypeError('Closed after abort request');
+
   stream._state = 'errored';
-  stream._storedError = new TypeError('Abort requested but closed successfully');
+  stream._storedError = error;
+
+  // TODO(tyoshino): No need to reject readyPromise?
 
-  WritableStreamRejectClosedPromiseIfAny(stream);
+  WritableStreamRejectClosedPromiseIfAny(stream, error);
 }
 
-function WritableStreamFinishPendingCloseWithError(stream, reason) {
-  assert(stream._pendingCloseRequest !== undefined);
-  stream._pendingCloseRequest._reject(reason);
-  stream._pendingCloseRequest = undefined;
+function WritableStreamFinishInFlightCloseWithError(stream, reason) {
+  assert(stream._inFlightCloseRequest !== undefined);
+  stream._inFlightCloseRequest._reject(reason);
+  stream._inFlightCloseRequest = undefined;
 
   const state = stream._state;
 
-  let wasAborted = false;
-  if (stream._pendingAbortRequest !== undefined) {
-    wasAborted = true;
-  }
+  if (state === 'errored') {
+    if (stream._pendingAbortRequest !== undefined) {
```

Comment:
I've factored out the code to handle the errored case.

I don't want to call WritableStreamRejectPromisesInReactionToError() in close() handling code just because it's harmless. It's weird to have the logic looks for stuff remaining in `[[writeRequests]]` and `[[closeRequest]]` after finishing close() processing.

Sometimes it helps to have all-in-one helper to take care of various situations. But for this case, I feel like that we shouldn't. Increasing call depth together with dependency from leaf function calls would sometimes lead to smaller but harder-to-read code.